File: Embarcadero/Berlin/Object Pascal/CopyPaste/fmMain.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit fmMain;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, FMX.TabControl, FMX.Objects, FMX.Controls.Presentation,
  FMX.StdCtrls, FMX.Edit, FMX.Colors;

type
  TForm3 = class(TForm)
    TabControl1: TTabControl;
    TabItem1: TTabItem;
    TabItem2: TTabItem;
    Panel1: TPanel;
    TextBorder: TRectangle;
    TextControl: TText;
    Label1: TLabel;
    Edit1: TEdit;
    TextRadioButton: TRadioButton;
    ImageRadioButton: TRadioButton;
    CopyButton: TButton;
    Panel2: TPanel;
    PasteImage: TImage;
    Button1: TButton;
    PasteLabel: TLabel;
    procedure Edit1ChangeTracking(Sender: TObject);
    procedure CopyButtonClick(Sender: TObject);
    procedure Button1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form3: TForm3;

implementation

{$R *.fmx}

uses
  System.Rtti, FMX.Platform, FMX.Surfaces;


procedure TForm3.Button1Click(Sender: TObject);
var
  Svc: IFMXClipboardService;
  Value: TValue;
  Bitmap: TBitmap;
begin
  if TPlatformServices.Current.SupportsPlatformService(IFMXClipboardService, Svc) then
  begin
    Value := Svc.GetClipboard;
    if not Value.IsEmpty then
    begin
      if Value.IsType<string> then
      begin
        PasteLabel.Text := Value.ToString;
        PasteImage.Bitmap := nil;
      end
      else if Value.IsType<TBitmapSurface> then
      try
        PasteLabel.Text := string.Empty;
        Bitmap := TBitmap.Create;
        try
          Bitmap.Assign(Value.AsType<TBitmapSurface>);
          PasteImage.Bitmap := Bitmap;
        finally
          Bitmap.Free;
        end;
      finally
        Value.AsType<TBitmapSurface>.Free;
      end;
    end;
  end;
end;

procedure TForm3.CopyButtonClick(Sender: TObject);
var
  Svc: IFMXClipboardService;
  Image: TBitmap;
begin
  if TPlatformServices.Current.SupportsPlatformService(IFMXClipboardService, Svc) then
    if TextRadioButton.IsChecked then
      Svc.SetClipboard(Edit1.Text)
    else
    begin
      Image := TextBorder.MakeScreenshot;
      try
        Svc.SetClipboard(Image);
      finally
        Image.Free;
      end;
    end;
end;

procedure TForm3.Edit1ChangeTracking(Sender: TObject);
begin
  CopyButton.Enabled := not Edit1.Text.IsEmpty;
  TextControl.Text := Edit1.Text;
end;

end.

File: Embarcadero/Berlin/CPP/CopyPaste/fmMain.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "fmMain.h"
#include <memory>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TForm4 *Form4;
//---------------------------------------------------------------------------
__fastcall TForm4::TForm4(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm4::CopyButtonClick(TObject *Sender)
{
	_di_IFMXClipboardService service;
	if(TPlatformServices::Current->SupportsPlatformService(__uuidof(IFMXClipboardService)) &&
		(service = TPlatformServices::Current->GetPlatformService(__uuidof(IFMXClipboardService)))) {
		if(TextRadioButton->IsChecked) {
			service->SetClipboard(TValue::_op_Implicit(Edit1->Text));
		}
		else {
			std::unique_ptr<TBitmap> Image(TextBorder->MakeScreenshot());
			service->SetClipboard(TValue::_op_Implicit(Image.get()));
        }
	}
}
//---------------------------------------------------------------------------
void __fastcall TForm4::Button1Click(TObject *Sender)
{
	_di_IFMXClipboardService service;
	if(TPlatformServices::Current->SupportsPlatformService(__uuidof(IFMXClipboardService)) &&
		(service = TPlatformServices::Current->GetPlatformService(__uuidof(IFMXClipboardService)))) {
		TValue value = service->GetClipboard();
		if(!value.IsEmpty) {
			if(value.IsType<String>()) {
				PasteLabel->Text = value.ToString();
				PasteImage->Bitmap = NULL;
			}
			else if (value.IsType<TBitmapSurface*>()) {
				PasteLabel->Text = EmptyStr;
				std::unique_ptr<TBitmap> bitmap(new TBitmap());
				bitmap->Assign(value.AsType<TBitmapSurface*>());
				PasteImage->Bitmap = bitmap.get();
				value.AsType<TBitmapSurface*>()->Free();
			}
		}
	}
}
//---------------------------------------------------------------------------
void __fastcall TForm4::Edit1ChangeTracking(TObject *Sender)
{
	CopyButton->Enabled = !Edit1->Text.IsEmpty();
	TextControl->Text = Edit1->Text;
}
//---------------------------------------------------------------------------
